### ISSUE TITLE
Support security mode TransportWithMessageCredential

### ISSUE THREAD

**Issue by iamjasonp** [dotnet/wcf#8]
Message level security is not currently supported.
Attempting to use it will throw PlatformNotSupportedException

**Comment by jeffboulanger** [comment]
Currently cannot move to our Windows 8.1 store application to Windows 10 because this is not implemented.

**Comment by oscar55** [comment]
Is there any progress related with this topic, Please i need it to work in it.

**Comment by ctolkien** [comment]
Just got burned by this one too.

**Comment by bc3tech** [comment]
wsHttpBinding, please.

**Comment by zhenlan** [comment]
This is currently still blocked by the lack of message security feature.

**Comment by haavardheggheim** [comment]
I need this too!

**Comment by ctolkien** [comment]
This has become more pressing for us with the announcement to ASP.NET Core 2 will not target NetStandard2. We need this in to move forward!

**Comment by Jelle87** [comment]
Any news wether this is on the short term roadmap?

**Comment by uwbadgerdev** [comment]
Any update on the status of fixing this?

**Comment by nickcoad** [comment]
This has been an issue now for 2 and a half years, why is this basic functionality still not supported?

**Comment by kevinciq** [comment]
I ended up sidestepping WCF altogether to solve this: https://stackoverflow.com/a/48818293

**Comment by OpenSpacesAndPlaces** [comment]
Any official workaround, or otherwise timeline for a patch? Thanks!

**Comment by OpenSpacesAndPlaces** [comment]
Normal Way
```
client = new MyClient(new BasicHttpBinding(BasicHttpSecurityMode.TransportWithMessageCredential), new EndpointAddress(BaseUrl));
client.ClientCredentials.UserName.UserName = UserName;
client.ClientCredentials.UserName.Password = Password;
```

Workaround
```
client = new MyClient(new BasicHttpBinding(BasicHttpSecurityMode.Transport), new EndpointAddress(BaseUrl));
client.Endpoint.SetMessageCredentials(UserName,Password);
```

Credit for solution basis:
https://stackoverflow.com/questions/734355/clueless-about-how-to-create-soap-wssesecurity-header
http://www.ideliverable.com/blog/calling-salesforce-soap-apis-from-a-net-core-or-uwp-app

```
	internal class AuthenticationHeader : MessageHeader
	{
		public AuthenticationHeader(string UserName, string Password,bool IncludeTimestamp)
		{
			_UserName = UserName;
			_Password = Password;
			_IncludeTimestamp = IncludeTimestamp;
		}

		private string _UserName;
		private string _Password;
		private bool _IncludeTimestamp;

		public override string Name => "Security";

		public override string Namespace => "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd";

		protected override void OnWriteStartHeader(System.Xml.XmlDictionaryWriter writer, System.ServiceModel.Channels.MessageVersion messageVersion)
		{
			writer.WriteStartElement("wsse", Name, Namespace);
			writer.WriteXmlnsAttribute("wsse", Namespace);
		}

		protected override void OnWriteHeaderContents(XmlDictionaryWriter writer, MessageVersion messageVersion)
		{
			//TODO Timestamp - If Needed

			writer.WriteStartElement("wsse", "UsernameToken", Namespace);

			writer.WriteStartElement("wsse", "Username", Namespace);
			writer.WriteValue(_UserName);
			writer.WriteEndElement();

			writer.WriteStartElement("wsse", "Password", Namespace);
			writer.WriteValue(_Password);
			writer.WriteEndElement();

			writer.WriteEndElement();
		}
	}
```

```
internal class AuthenticationHeaderBehavior : IEndpointBehavior
	{
		public AuthenticationHeaderBehavior(string UserName, string Password, bool IncludeTimestamp)
		{
			_UserName = UserName;
			_Password = Password;
			_IncludeTimestamp = IncludeTimestamp;
		}

		private string _UserName;
		private string _Password;
		private bool _IncludeTimestamp;

		public void AddBindingParameters(ServiceEndpoint endpoint, BindingParameterCollection bindingParameters)
		{

		}

		public void ApplyClientBehavior(ServiceEndpoint endpoint, ClientRuntime clientRuntime)
		{
			clientRuntime.ClientMessageInspectors.Add(new AuthenticationHeaderInspector(_UserName, _Password, _IncludeTimestamp));
		}

		public void ApplyDispatchBehavior(ServiceEndpoint endpoint, EndpointDispatcher endpointDispatcher)
		{

		}

		public void Validate(ServiceEndpoint endpoint)
		{

		}
	}
```
```
	internal static class AuthenticationHeaderExtensions
	{
		public static void SetMessageCredentials(this ServiceEndpoint endpoint, string UserName, string Password)//, bool IncludeTimestamp)
		{
			//TODO Timestamp - If Needed
			endpoint.EndpointBehaviors.Add(new AuthenticationHeaderBehavior(UserName, Password, false));//IncludeTimestamp));
		}
	}
```
```
	internal class AuthenticationHeaderInspector : IClientMessageInspector
	{
		public AuthenticationHeaderInspector(string UserName, string Password, bool IncludeTimestamp)
		{
			_UserName = UserName;
			_Password = Password;
			_IncludeTimestamp = IncludeTimestamp;
		}

		private string _UserName;
		private string _Password;
		private bool _IncludeTimestamp;

		public void AfterReceiveReply(ref Message reply, object correlationState)
		{

		}

		public object BeforeSendRequest(ref Message request, IClientChannel channel)
		{
			request.Headers.Add(new AuthenticationHeader(_UserName, _Password, _IncludeTimestamp));

			return Guid.NewGuid();
		}
	}
```

**Comment by Jhatzig** [comment]
I try your solution but my result is : 
MessageSecurityException: The HTTP request is unauthorized with client authentication scheme 'Basic'. The authentication header received from the server was 'Basic Realm = 'Integration Server''. :(

**Comment by OpenSpacesAndPlaces** [comment]
@Jhatzig 

I'm not sure on your specific case - but I'd confirm you have the correct "BasicHttpSecurityMode" set.
In my build out it went from:
`BasicHttpSecurityMode.TransportWithMessageCredential`
to:
`BasicHttpSecurityMode.Transport`

Also keep in mind that not all SOAP authentication security headers are the same.
You may need to customize the XML generated by:
` AuthenticationHeader : MessageHeader`

The output this generates is like:
```
  <soap:Header>
    <wsse:Security xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd">
      <wsse:UsernameToken>
        <wsse:Username>user</wsse:Username>
        <wsse:Password>password</wsse:Password>
      </wsse:UsernameToken>
    </wsse:Security>
  </soap:Header>
```

**Comment by Jhatzig** [comment]
@OpenSpacesAndPlaces 

Thanks your for your answer.
My case is : 
A WS with WCF in HTTP with **TransportCredentialOnly**
With .NET classic, in web.config , on binding, we add the message security : 
 ` <message clientCredentialType="UserName" algorithmSuite="Default" />`

but with .Net Core, we can't do that. 
So, I try to use your solution but it's for **TransportWithMessageCredential** and i'm not founding the solution for *TransportCredentialOnly**

If, i'm setting to "Transport" i have the following error : 
ArgumentException: The provided URI scheme 'http' is invalid; expected 'https'.
because my WS is in HTTP

**Comment by OpenSpacesAndPlaces** [comment]
@Jhatzig 

`BasicHttpSecurityMode.Transport`
Or similar, is going to require an "Https" url.

You may be looking for (but I'd caution using that in anything but a development environment):
`BasicHttpSecurityMode.None`

As far as using the equivalent of:
`TransportCredentialOnly`

To my understanding, you'd need HTTP headers instead of a SOAP header - so something like this (untested) - is probably what you're after. That's kind of the direction you'll need to dig is how to add HTTP headers to a WCF requests manually.

`BasicHttpSecurityMode.None`

```
	internal class AuthenticationHeaderInspector : IClientMessageInspector
	{
		public AuthenticationHeaderInspector(string UserName, string Password, bool IncludeTimestamp)
		{
			_UserName = UserName;
			_Password = Password;
			_IncludeTimestamp = IncludeTimestamp;
		}

		private string _UserName;
		private string _Password;
		private bool _IncludeTimestamp;

		public void AfterReceiveReply(ref Message reply, object correlationState)
		{

		}

		public object BeforeSendRequest(ref Message request, IClientChannel channel)
		{
			using (new System.ServiceModel.OperationContextScope(channel))
			{
				MessageHeader head = MessageHeader.CreateHeader("Authorization", "http://yournamespace.com/v1", "Basic yourcredentials");
				OperationContext.Current.OutgoingMessageHeaders.Add(head);
			}

			return Guid.NewGuid();
		}
```

**Comment by mconnew** [comment]
While this works, be aware that it doesn't support WS-SecureConversation which is needed for default support for WSHttpBinding. The difference when using WS-SecureConversation is a handshake happens at channel open where a secure token is negotiated and used in actual payload messages being sent back and forth. This has two advantages. First, it allows a persistent session on the server side so you can have a persistent instance context/service instance for a single client channel. Second, with more complicated tokens than username and password, the token validation can be a bit more expensive. Sending the full authentication token every time requires it to be re-validated for every request.  The problem we have with in-box support is WCF depends on System.IdentityModel which the owning team don't wish to port to .NET Core. So far I've prototyped on the .NET Framework removing most of the crypto dependency of IdentityModel and I was able to get this scenario working with minimal dependency on IdentityModel (using the SignedXml implementation that's available in .NET Core instead). The next step is to bring over the WCF pieces and my prototype to .NET Core along with some pieces of IdentityModel which we will now need to own.  
So in short, there are some difficult blockers but I'm working on it and have a path to get this working with a successful proof of concept prototype.

**Comment by elmojo** [comment]
It's been three and a half years and this issue still hasn't been resolved?  I too am being bit by this issue.  I will try the work-arounds suggested above but it's kind of ridiculous that this has not been fixed yet.

**Comment by elmojo** [comment]
Yes it did... BUT...

Ugh, this fixed the login issue but I then received a new error.  Let me explain.

Trying to use WCF within .NET Core has been a frustrating experience.  I would have been much better off just using a HttpClient and parsing the XML manually.  I've spent three days messing with this and still don't have a working solution.

It has been a never ending game of whack-a-mole.  Fix one error, BAM, up pops another.  Fix that one, BAM, another.  You get the picture.

So here are the multitude of problems I've experienced using WCF with .Net Core.

1.  I plugged the WSDL into the SvcUtil and chose to use generic Lists instead of arrays.  Write some code to test it out, BAM, error:  System.PlatformNotSupportedException: Compiling JScript/CSharp scripts is not supported.  Search the internet for a solution... no solution still an open [issue](https://github.com/dotnet/wcf/issues/1274) that's over two years old.
2.  Implement the workaround described in the issue above.  Rerun test.  BAM, error: The remote certificate is invalid according to the validation procedure.  Back to the internet... turns out that the SSL certificate IS VALID it's just that because it's just a development box, the certificate is self-signed.  So the error reported isn't very helpful or correct.
3.  Implement a workaround to accept the certificate.  Rerun test.  BAM, error: System.ServiceModel.Security.SecurityAccessDeniedException : The security token could not be authenticated or authorized.  Back to the internet...
4.  Turns out I need to implement TransportWithMessageCredential on the binding.  Ok, change that bit of code, rerun test.  BAM, error: System.PlatformNotSupportedException : The value 'TransportWithMessageCredential' is not supported in this context for the binding security property 'value'.  Turns out this feature hasn't been implemented yet which lead me to this thread.  ISSUE number freaking #8 on the repository.  The repo has had over 3000 issues documented but #8 is still outstanding?
5.  Fine, I implement @OpenSpacesAndPlaces work around.  Had to tweak a few things so it'd match my schema.  Rerun test.  BAM, error: System.ServiceModel.ProtocolException : The content type text/xml; charset=us-ascii of the response message does not match the content type of the binding (text/xml; charset=utf-8). If using a custom encoder, be sure that the IsContentTypeSupported method is implemented properly.  UGH!!!!!!!!
6.  Ok, I'll just set the binding to ASCII.  Rerun test.  BAM, error: System.ArgumentException : The text encoding 'us-ascii' used in the text message format is not supported.Parameter name: encoding.  SHOOT ME NOW!  Yet one more feature not yet implemented by WCF.

So now my options are to try and convince the service owner to spit out UTF-8 instead of ASCII, write a custom binding solution, or just use HttpClient and manually parse the XML response.  If I can't convince them to switch the character set (and even if they do, does another mole pop-up for me to have to whack), I probably go with HttpClient which I can probably write quicker than a custom binding (which  if I do still does not guarantee a working solution, those moles are lurking).

**Comment by OpenSpacesAndPlaces** [comment]
@elmojo 

I haven't tested it, but this issue thread:
https://github.com/dotnet/wcf/issues/1965

Suggests you should be able work around with a customer encoder like:
https://github.com/dotnet/wcf/blob/master/src/System.Private.ServiceModel/tools/IISHostedWcfService/App_code/testhosts/CustomTextMessageEncoder.cs

I'm not sure on your exact case, but it's plausible to me you might have to not specify an encoding (or use one it 'happy' with) and then wire in the work around.

e.g. like the sample for transport security needing to drop:
`BasicHttpSecurityMode.TransportWithMessageCredential`
to:
`BasicHttpSecurityMode.Transport`

Hope this helps.

**Comment by elmojo** [comment]
@OpenSpacesAndPlaces 

Thanks for that, I'll check it out.  We have a request in from the service provider on whether they can switch the character set.  If they won't, I'll try the code in the link you provided.

**Comment by elmojo** [comment]
@OpenSpacesAndPlaces 

So the service owner agreed to create a new service that output UTF-8.

7.  Point my test to the new service.  BAM, error: System.ServiceModel.QuotaExceededException : The maximum message size quota for incoming messages (65536) has been exceeded. To increase the quota, use the MaxReceivedMessageSize property on the appropriate binding element.  Seems like an easy fix.  I increase the MaxReceivedMessageSize property on my binding to a sufficiently large number (the property's type is 'long').
8.  Rerun test.  BAM, error: System.ArgumentOutOfRangeException : This factory buffers messages, so the message sizes must be in the range of an integer value.  So even though the property is of type 'long' something downstream expects an 'int'.  So I set MaxReceivedMessageSize to the max for a signed int.
9.  Rerun.  BAM, error: System.InvalidOperationException : The username is not provided. Specify username in ClientCredentials.  So now along with the workaround that @OpenSpacesAndPlaces provided, I go ahead and add the credentials to the service's ClientCredentials.  This is originally how I had it configured but had taken it out with the workaround (duh, transport with message).
10.  Rerun.  SUCCESS!  Wow, it actually worked without errors.  Debug, inspect the returned parsed object... FAIL!  It didn't actually serialize anything into the data object.  It appears to have gotten the record count correct, but other than that, the data object is empty.  All of this work for nothing.

At this point, I'm giving up on WCF for this particular solution.  For something that was supposed to save me time, it's done anything but.

**Comment by mwrothbe** [comment]
@mconnew Is there a solution for WSHttpBinding?  As you stated, the solutions posted here aren't working for this scenario.

**Comment by dgroh** [comment]
This is also not working:

```
            var myBinding = new BasicHttpsBinding();
            myBinding.Security.Transport.ClientCredentialType = HttpClientCredentialType.None;
            myBinding.Security.Mode = BasicHttpsSecurityMode.Transport;
```

This is still an issue? That is impossible...

**Comment by tmulkern** [comment]
Hi All, I know I am late to the conversation, but I wrote and article that might help with this: 
[Using a Bearer Token in .Net Core with WCF](https://medium.com/@tadhg.j.mulkern/using-a-bearer-token-in-net-core-with-wcf-part-1-8db2c54113f8)

**Comment by rpj** [comment]
@Lxiamail: I noticed that you recently added the "1 - Up Next" label to this issue, does that mean that it is in the queue to be implemented soon? That would be very exciting! If so, do you have a timeline as to when we could expect this feature to be available in a proper .NET core release? Thank you!

**Comment by Lxiamail** [comment]
@rpj We are working on a prototype of TransportWithMessageCredential. However, due to another competing priority work item popped up recently, this feature's timeline is still to be determined.

**Comment by rpj** [comment]
@Lxiamail Thank you so much for the information Alicia, it is very much appreciated!

**Comment by Amikumar** [comment]
can anyone help me to provide a workaround to fix this issue ?

**Comment by lmcarreiro** [comment]
@Amikumar I used this workaround:

```cs
        private Workforce_2_3_0PTClient GetClient()
        {
            var address = new System.ServiceModel.EndpointAddress(endpointUri);
            var binding = new System.ServiceModel.BasicHttpBinding()
            {
                MaxBufferSize = int.MaxValue,
                MaxBufferPoolSize = int.MaxValue,
                MaxReceivedMessageSize = int.MaxValue,
            };

            var client = new Workforce_2_3_0PTClient(binding, address);
            client.Endpoint.EndpointBehaviors.Add(new BasicAuthenticationEndpointBehavior(usernameOsb, passwordOsb));

            return client;
        }

        private class BasicAuthenticationEndpointBehavior : System.ServiceModel.Description.IEndpointBehavior
        {
            private readonly string basicAuthenticationData;

            public BasicAuthenticationEndpointBehavior(string username, string password)
            {
                var bytes = System.Text.Encoding.UTF8.GetBytes($"{username}:{password}");
                basicAuthenticationData = Convert.ToBase64String(bytes);
            }

            void System.ServiceModel.Description.IEndpointBehavior.AddBindingParameters(System.ServiceModel.Description.ServiceEndpoint endpoint, System.ServiceModel.Channels.BindingParameterCollection bindingParameters)
            {

            }

            void System.ServiceModel.Description.IEndpointBehavior.ApplyClientBehavior(System.ServiceModel.Description.ServiceEndpoint endpoint, System.ServiceModel.Dispatcher.ClientRuntime clientRuntime)
            {
                BasicAuthenticationMessageInspector inspector = new BasicAuthenticationMessageInspector(basicAuthenticationData);
                clientRuntime.ClientMessageInspectors.Add(inspector);
            }
            
            void System.ServiceModel.Description.IEndpointBehavior.ApplyDispatchBehavior(System.ServiceModel.Description.ServiceEndpoint endpoint, System.ServiceModel.Dispatcher.EndpointDispatcher endpointDispatcher)
            {

            }
            
            void System.ServiceModel.Description.IEndpointBehavior.Validate(System.ServiceModel.Description.ServiceEndpoint endpoint)
            {

            }
        }

        private class BasicAuthenticationMessageInspector : System.ServiceModel.Dispatcher.IClientMessageInspector
        {
            private const string HEADER_NAME = "Authorization";
            private readonly string authenticationData;

            public BasicAuthenticationMessageInspector(string authenticationData)
            {
                this.authenticationData = authenticationData;
            }

            object System.ServiceModel.Dispatcher.IClientMessageInspector.BeforeSendRequest(ref System.ServiceModel.Channels.Message request, System.ServiceModel.IClientChannel channel)
            {
                if (request.Properties.TryGetValue(System.ServiceModel.Channels.HttpRequestMessageProperty.Name, out object httpRequestMessageObject)
                    && httpRequestMessageObject is System.ServiceModel.Channels.HttpRequestMessageProperty httpRequestMessage)
                {
                    if (string.IsNullOrEmpty(httpRequestMessage.Headers[HEADER_NAME]))
                    {
                        httpRequestMessage.Headers[HEADER_NAME] = $"Basic {authenticationData}";
                    }
                }
                else
                {
                    httpRequestMessage = new System.ServiceModel.Channels.HttpRequestMessageProperty();
                    httpRequestMessage.Headers.Add(HEADER_NAME, $"Basic {authenticationData}");
                    request.Properties.Add(System.ServiceModel.Channels.HttpRequestMessageProperty.Name, httpRequestMessage);
                }
                return null;
            }

            void System.ServiceModel.Dispatcher.IClientMessageInspector.AfterReceiveReply(ref System.ServiceModel.Channels.Message reply, object correlationState)
            {

            }
        }
```

**Comment by Lxiamail** [comment]
@mconnew Can you move out the SecurityModel Message support work item to future milestone since it ais blocked by MessageSecurity not available on .NET Core.

**Comment by mconnew** [comment]
This issue has been changed to only be for tracking the security mode TransportWithMessageCredential. Full Message security will be tracked in issue #3424.

**Comment by mconnew** [comment]
As anything related to message security has so many options and flavors that to bring everything over will be too much work (as some re-writing is needed) to do all at once so the approach I'll be taking will be scenario based. Initially I will be implementing UserName (see #1257) and X509Certificate credential tokens.

**Comment by TomProcureur** [comment]
> Normal Way
> 
> ```
> client = new MyClient(new BasicHttpBinding(BasicHttpSecurityMode.TransportWithMessageCredential), new EndpointAddress(BaseUrl));
> client.ClientCredentials.UserName.UserName = UserName;
> client.ClientCredentials.UserName.Password = Password;
> ```
> Workaround
> 
> ```
> client = new MyClient(new BasicHttpBinding(BasicHttpSecurityMode.Transport), new EndpointAddress(BaseUrl));
> client.Endpoint.SetMessageCredentials(UserName,Password);
> ```
> Generates a header like:
> 
> ```
>  <soap:Header>
>     <wsse:Security xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd">
>       <wsse:UsernameToken>
>         <wsse:Username>user</wsse:Username>
>         <wsse:Password>password</wsse:Password>
>       </wsse:UsernameToken>
>     </wsse:Security>
>   </soap:Header>
> ```
> Credit for solution basis:
> https://stackoverflow.com/questions/734355/clueless-about-how-to-create-soap-wssesecurity-header
> http://www.ideliverable.com/blog/calling-salesforce-soap-apis-from-a-net-core-or-uwp-app
> 
> ```
> 	internal class AuthenticationHeader : MessageHeader
> 	{
> 		public AuthenticationHeader(string UserName, string Password,bool IncludeTimestamp)
> 		{
> 			_UserName = UserName;
> 			_Password = Password;
> 			_IncludeTimestamp = IncludeTimestamp;
> 		}
> 
> 		private string _UserName;
> 		private string _Password;
> 		private bool _IncludeTimestamp;
> 
> 		public override string Name => "Security";
> 
> 		public override string Namespace => "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd";
> 
> 		protected override void OnWriteStartHeader(System.Xml.XmlDictionaryWriter writer, System.ServiceModel.Channels.MessageVersion messageVersion)
> 		{
> 			writer.WriteStartElement("wsse", Name, Namespace);
> 			writer.WriteXmlnsAttribute("wsse", Namespace);
> 		}
> 
> 		protected override void OnWriteHeaderContents(XmlDictionaryWriter writer, MessageVersion messageVersion)
> 		{
> 			//TODO Timestamp - If Needed
> 
> 			writer.WriteStartElement("wsse", "UsernameToken", Namespace);
> 
> 			writer.WriteStartElement("wsse", "Username", Namespace);
> 			writer.WriteValue(_UserName);
> 			writer.WriteEndElement();
> 
> 			writer.WriteStartElement("wsse", "Password", Namespace);
> 			writer.WriteValue(_Password);
> 			writer.WriteEndElement();
> 
> 			writer.WriteEndElement();
> 		}
> 	}
> ```
> ```
> internal class AuthenticationHeaderBehavior : IEndpointBehavior
> 	{
> 		public AuthenticationHeaderBehavior(string UserName, string Password, bool IncludeTimestamp)
> 		{
> 			_UserName = UserName;
> 			_Password = Password;
> 			_IncludeTimestamp = IncludeTimestamp;
> 		}
> 
> 		private string _UserName;
> 		private string _Password;
> 		private bool _IncludeTimestamp;
> 
> 		public void AddBindingParameters(ServiceEndpoint endpoint, BindingParameterCollection bindingParameters)
> 		{
> 
> 		}
> 
> 		public void ApplyClientBehavior(ServiceEndpoint endpoint, ClientRuntime clientRuntime)
> 		{
> 			clientRuntime.ClientMessageInspectors.Add(new AuthenticationHeaderInspector(_UserName, _Password, _IncludeTimestamp));
> 		}
> 
> 		public void ApplyDispatchBehavior(ServiceEndpoint endpoint, EndpointDispatcher endpointDispatcher)
> 		{
> 
> 		}
> 
> 		public void Validate(ServiceEndpoint endpoint)
> 		{
> 
> 		}
> 	}
> ```
> ```
> 	internal static class AuthenticationHeaderExtensions
> 	{
> 		public static void SetMessageCredentials(this ServiceEndpoint endpoint, string UserName, string Password)//, bool IncludeTimestamp)
> 		{
> 			//TODO Timestamp - If Needed
> 			endpoint.EndpointBehaviors.Add(new AuthenticationHeaderBehavior(UserName, Password, false));//IncludeTimestamp));
> 		}
> 	}
> ```
> ```
> 	internal class AuthenticationHeaderInspector : IClientMessageInspector
> 	{
> 		public AuthenticationHeaderInspector(string UserName, string Password, bool IncludeTimestamp)
> 		{
> 			_UserName = UserName;
> 			_Password = Password;
> 			_IncludeTimestamp = IncludeTimestamp;
> 		}
> 
> 		private string _UserName;
> 		private string _Password;
> 		private bool _IncludeTimestamp;
> 
> 		public void AfterReceiveReply(ref Message reply, object correlationState)
> 		{
> 
> 		}
> 
> 		public object BeforeSendRequest(ref Message request, IClientChannel channel)
> 		{
> 			request.Headers.Add(new AuthenticationHeader(_UserName, _Password, _IncludeTimestamp));
> 
> 			return Guid.NewGuid();
> 		}
> 	}
> ```

When i use this approach, i got the error:
FaultException: Missing Timestamp
System.ServiceModel.Channels.ServiceChannel.HandleReply(ProxyOperationRuntime operation, ref ProxyRpc rpc)

How could i solve this?

**Comment by lmcarreiro** [comment]
@TomProcureur have you tried to use [this workaround](https://github.com/dotnet/wcf/issues/8#issuecomment-464888793) that puts the credentials in the HTTP request header instead of the SOAP message header?

**Comment by OpenSpacesAndPlaces** [comment]
@TomProcureur 

If you look closely there is a comment in the code:
`		//TODO Timestamp - If Needed`

The particular provider I was working with didn't need that.

You'll need to add the xml for:
`wsu:Timestamp`

**Comment by OpenSpacesAndPlaces** [comment]
@TomProcureur 

You'll need to adjust the code to your specific use case - but here is a sample you can work from (untested):
```
 writer.WriteStartElement("wsu", "Timestamp", wsuNamespace);
            writer.WriteAttributeString("Id", "TS-" + Guid.NewGuid().ToString());
            writer.WriteStartElement("wsu", "Created");
            writer.WriteValue(_createdDate.ToString("yyyy-MM-ddThh:mm:ss"));
            writer.WriteEndElement();

            writer.WriteStartElement("wsu", "Expires");
            writer.WriteValue(_createdDate.AddMinutes(20).ToString("yyyy-MM-ddThh:mm:ss"));
            writer.WriteEndElement();
```

Pulled from:
https://social.msdn.microsoft.com/Forums/vstudio/en-US/fb0970a6-eb5d-44ec-8060-90d0a3cc76a1/adding-soap-header-in-an-iclientmessageinspector-working-fine-with-one-root-element-but-how-do-i?forum=wcf

**Comment by TomProcureur** [comment]
@OpenSpacesAndPlaces 
I've changed the code to include a timestamp:

` protected override void OnWriteHeaderContents(XmlDictionaryWriter writer, MessageVersion messageVersion)
        {

            var createdDate = DateTime.Now;

            writer.WriteStartElement("wsu", "Timestamp", "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd");
            writer.WriteAttributeString("wsu", "Id", "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd", "TS-" + Guid.NewGuid().ToString("N").ToUpper());

            writer.WriteStartElement("wsu", "Created", "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd");
            writer.WriteValue(createdDate.ToString("yyyy-MM-ddTHH:mm:ss.fffZ"));
            writer.WriteEndElement();

            writer.WriteStartElement("wsu", "Expires", "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd");
            writer.WriteValue(createdDate.AddMinutes(5000).ToString("yyyy-MM-ddTHH:mm:ss.fffZ"));
            writer.WriteEndElement();

            writer.WriteEndElement();

            writer.WriteStartElement("wsse", "UsernameToken", Namespace);

            writer.WriteStartElement("wsse", "Username", Namespace);
            writer.WriteValue(_UserName);
            writer.WriteEndElement();

            writer.WriteStartElement("wsse", "Password", Namespace);
            writer.WriteValue(_Password);
            writer.WriteEndElement();

            writer.WriteEndElement();
        }`

But now this gives me the error:

System.ServiceModel.FaultException: The timestamp could not be validated
   at System.ServiceModel.Channels.ServiceChannel.HandleReply(ProxyOperationRuntime operation, ProxyRpc& rpc)
   at System.ServiceModel.Channels.ServiceChannel.EndCall(String action, Object[] outs, IAsyncResult result)
   at System.ServiceModel.Channels.ServiceChannelProxy.TaskCreator.<>c__DisplayClass1_0.<CreateGenericTask>b__0(IAsyncResult asyncResult)

Any thoughts on how this could be solved? 

When i check the log files on the server running the service, everythings get processed correctly and a response is send back to the client. 

But inside the client i get the 'The timestamp could not be validated' instead of the correct response

**Comment by OpenSpacesAndPlaces** [comment]
@TomProcureur 

Unfortunately I'm not readily sure to the specific cause.

I'd tried to see if you can get the raw response from `AfterReceiveReply` or if not there then on the sending side - just to see if anything jumps out.

-Once obvious thing to check would be that the timestamps you are sending/receiving are valid:
5000 minutes may be too far into the future - so maybe start with 5 minutes and work up.
-Also double check the the timezone between the request and the machine is in agreement.
-More remotely, if the datetime is totally out of sync between the machines, that could be it too.

**Comment by TomProcureur** [comment]
@OpenSpacesAndPlaces 

Allright, i was using the wrong timezone when creating my timestamp.
I now use the DateTime.UtcNow and this works better...

But there is still an issue. Now i get the error:

_An unhandled exception occurred while processing the request.
The header 'Security' from the namespace 'http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd' was not understood by the recipient of this message, causing the message to not be processed.  This error typically indicates that the sender of this message has enabled a communication protocol that the receiver cannot process.  Please ensure that the configuration of the client's binding is consistent with the service's binding._ 


But when I inspect the raw response inside the AfterReceiveReply everything looks correct and i see the correct result:

`?xml version="1.0" encoding="utf-16"?>
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
  <soapenv:Header>
    <wsse:Security xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" soapenv:mustUnderstand="1">
      <wsu:Timestamp xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" wsu:Id="Timestamp-56542">
        <wsu:Created>2019-03-20T08:15:25.709Z</wsu:Created>
        <wsu:Expires>2019-03-20T08:20:25.709Z</wsu:Expires>
      </wsu:Timestamp>
    </wsse:Security>
  </soapenv:Header>
  <soapenv:Body>
         // ...
  </soapenv:Body>
</soapenv:Envelope>`

What could that be?

**Comment by TomProcureur** [comment]
I've fixed this by modifying the security header from the reply

` public void AfterReceiveReply(ref Message reply, object correlationState)
        {
            reply = FixMessageHeader(reply);
        }`

`private Message FixMessageHeader(Message oldMessage)
        {
            var ms = new MemoryStream();
            XmlWriter writer = XmlWriter.Create(ms);
            oldMessage.WriteMessage(writer); // the message was consumed here
            writer.Flush();
            ms.Position = 0;
            XmlDocument xmlDoc = new XmlDocument();
            xmlDoc.Load(ms);


            //Modify mustUnderstand attribute
            XmlNamespaceManager nsmgr = new XmlNamespaceManager(xmlDoc.NameTable);
            nsmgr.AddNamespace("soapenv", "http://schemas.xmlsoap.org/soap/envelope/");
            nsmgr.AddNamespace("wsse", "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd");

            XmlNode node = xmlDoc.SelectSingleNode("//soapenv:Envelope/soapenv:Header/wsse:Security", nsmgr);
            if (node.Attributes["soapenv:mustUnderstand"].Value.ToString() == "1")
                node.Attributes["soapenv:mustUnderstand"].Value = "0";


            //Now recreating the message
            ms = new MemoryStream();
            xmlDoc.Save(ms);
            ms.Position = 0;
            XmlReader reader = XmlReader.Create(ms);

            Message newMessage = Message.CreateMessage(reader, int.MaxValue, oldMessage.Version);
            newMessage.Properties.CopyProperties(oldMessage.Properties);
            return newMessage;
        }`

**Comment by mconnew** [comment]
I now have TransportWithMessageCredentials working using certificates. It still needs a lot of cleanup work so I won't be merging it immediately but if anyone wants to build WCF themselves and try it out, you can find it on my branch [here](https://github.com/mconnew/wcf/tree/8-TransportWithMessageCredential). I would expect username credentials to work but I haven't tested it yet.

**Comment by mconnew** [comment]
Username credentials didn't work at the time, but they do now in the current pull request.

**Comment by memuyskens** [comment]
I downloaded the latest version of your branch, built it, and added the .dlls to my solution. However, it seems that only WSHttpBinding supports messages and unfortunately the legacy service I'm forced to use requires BasicHttpBinding with TransportWithMessageCredential. Will this be supported in the future?

**Comment by mconnew** [comment]
@memuyskens, it wasn't initially intended to be as it wasn't clear that was being asked for but it's only a small addition to add that. You could probably get it working using a custom binding, but support with BasitHttpBinding should be there. Can you provide your binding details so I can write a test scenario to make sure your use case is covered?

**Comment by memuyskens** [comment]
I'm new to soap, so I may not provide everything here you need, but this is what I know:
Mode: TransportWithMessageCredential
Transport: client credential type & proxy credential type are both 'None'
Message: client credential type is 'Certificate' and algo suite is 'Default'
Hope this helps!
I'll have to look into custom bindings to solve this issue; if you have any pointers that'd be great. Thanks

**Comment by kayakingcoder** [comment]
I also require BasicHttpBinding to talk to an older version of SAP.

Here are my binding details:

```
                    binding = new BasicHttpBinding();
                    binding.Security.Mode = BasicHttpSecurityMode.Transport;
                    binding.Security.Transport.ClientCredentialType = HttpClientCredentialType.Basic;
                    binding.MaxReceivedMessageSize = 2097152;
                    binding.SendTimeout = new TimeSpan(0, 3, 0);
                    binding.ReceiveTimeout = new TimeSpan(0, 3, 0);

```

**Comment by mconnew** [comment]
@kayakingcoder, your scenario is already supported. This issue is for BasicHttpSecurityMode.TransportWithMessageCredential

**Comment by mconnew** [comment]
@memuyskens, if you try my branch again, you will find it works for BasicHttpBinding now.

**Comment by mconnew** [comment]
WSHttpBinding and BasicHttpBinding support for TransportWithMessageCredential  with Username and Certificate credentials now merged to master. Next up, support for NetTcpBinding. We have some problems with publishing development packages right now so if you want to test, you will need to build the code yourself until that's resolved.  
I would appreciate everyone testing as soon as possible as this area of functionality has many knobs and I only implemented the ones necessary for specific scenarios. If your scenario isn't supported yet, mentioning it early will increase the chance of it being fixed before 3.0 release.

**Comment by crodeheaver** [comment]
Are there any plans to support WS2007FederationHttpBinding?

**Comment by AliNazar1991** [comment]
> WSHttpBinding and BasicHttpBinding support for TransportWithMessageCredential with Username and Certificate credentials now merged to master. Next up, support for NetTcpBinding. We have some problems with publishing development packages right now so if you want to test, you will need to build the code yourself until that's resolved.
> I would appreciate everyone testing as soon as possible as this area of functionality has many knobs and I only implemented the ones necessary for specific scenarios. If your scenario isn't supported yet, mentioning it early will increase the chance of it being fixed before 3.0 release.

In which release of .net core 3.0 will these that you have made be available? I have tried with .net core 3 preview 5 but i still get:
The value 'TransportWithMessageCredential' is not supported in this context for the binding security property 'securityMode'.

**Comment by mconnew** [comment]
@AliNazar1991, I have a PR open right now with changes to implement it for NetTcp. We have a deadlock somewhere in the new codebase that is only showing up on our test infrastructure that I need to resolve before I can merge. But if you want to clone that PR locally and build it, it should work. The deadlock is only related to using a non-default synchronization context, and specifically xunit won't complete the test unless all operations going through the sync context that are started are also completed. If that doesn't mean anything to you, don't worry. It's an edge case that needs to be fixed before RTM but won't affect 99.999% of people. If you are just looking for timing, I would expect it to be in the next preview release.

**Comment by cristicatalan** [comment]
I can see secure WSHttpBinding with TransportWithMessageCredential is on the way to .Net Core 3.0. Will there also be support for consuming WCF services having wsHttpContextBinding in .Net Core 3.0?

If service Web.config has endpoint such as:
`<endpoint address="" binding="**wsHttpContextBinding**" bindingConfiguration="**MyService_Secured**" contract="IMyService" />`

With binding:
``` 
   <bindings>      
      <**wsHttpContextBinding**>
        <binding name="**MyService_Secured**" ...>
          <readerQuotas ... />
          <security mode="TransportWithMessageCredential">
            <transport clientCredentialType="None" />
            <message clientCredentialType="UserName" establishSecurityContext="false" />
          </security>
        </binding>
      </**wsHttpContextBinding**>
   </binding>
```

**Comment by mconnew** [comment]
`WSHttpContextBinding` won't be in .NET Core 3.0. The context feature is a whole other set of classes that would need to be ported in addition to `WSHttpBinding`. The context feature was designed to be used with Workflow to enable connecting to a Workflow service host and picking up where you were previously for long running workflows. Are you using this outside of Workflow? If so, could you describe your scenario briefly? To be honest with you, this is the first time I've even heard of someone actually using the feature.

**Comment by cristicatalan** [comment]
I don't use Workflow. Thanks for this clarification. The project is part of some legacy code and I suspect `WSHttpContextBinding` was used unintentionally instead of `WSHttpBinding`.
I changed the WCF service configuration to use `WSHttpBinding`, downloaded this repo code, compiled it locally and referenced it in my .NET Core project and it works fine.

**Comment by mconnew** [comment]
NetTcp, BasicHttp and WSHttpBinding now support TransportWithMessageCredential checked in to master. I would appreciate if anyone needing this feature could try it out for their scenario as it would be best to know what sub-features people need that aren't there sooner rather than later.

**Comment by sutaro** [comment]
Will TransportWithMessageCredential be available in .net core 2?

**Comment by mconnew** [comment]
We are not moving our dependencies forward to require .NET Core 3.0 as there are no new api's that we need. If you want to stay on .NET Core 2, because of changes to HttpClient, I would recommend using .NET Core 2.2. We still only require netstandard 2.0 so we should still work back to .NET Core 2.0 but there's some behavior differences with HTTP client certificate authentication which might cause you problems if you are using client certs. In earlier releases we have a best effort mitigation because of lack of Continue-100 support in earlier versions of .NET Core which we've now removed. It's only a possible, so you might be fine. If you aren't using client certs, it just becomes a performance issue.

**Comment by CumpsD** [comment]
@mconnew Any feedback on @crodeheaver question about supporting `WS2007FederationHttpBinding`?

**Comment by mconnew** [comment]
@CumpsD, we're in the process of adding support for WS-Federation. We're waiting on another team to provide an implementation of a WS-Trust client which can get the federation token to then communicate with the server. That's then used by WSHttpBinding to communicate with the server. I took a look at WS2007FederationHttpBinding to see how it's different than WSFederationHttpBinding and the difference is that it uses WS2007HttpBinding which we don't have as a specific binding. I checked the protocols used by WS2007HttpBinding and it looks like we probably have most if not all the implementation required to implement it. Can you open an issue specifically for WS2007FederationHttpBinding?

**Comment by CumpsD** [comment]
Thanks for the quick follow up @mconnew! I've gone ahead and opened a new issue to track it. I'll let it up to you to properly label it according to the repo's standards.

**Comment by nmarun** [comment]
Hi,

I'm trying to connect to a WCF service which has the below settings:

`<binding name="GDASServFY14R22binding" hostNameComparisonMode="StrongWildcard" rec...
  <readerQuotas maxDepth="2147483647" maxStringContentLength="2147483647" maxArray...
  <security mode="TransportCredentialOnly">
    <transport clientCredentialType="Ntlm" proxyCredentialType="None" realm="" />
    <message clientCredentialType="UserName" algorithmSuite="Default" />
  </security>
`
My API is in .net core 2.2. I'm not seeing Default as the algorithm suite. Is this not supported? If so, is there a workaround?

Thanks,
Arun

**Comment by OpenSpacesAndPlaces** [comment]
@nmarun 

I'm not 100% sure on your specific use case, but for some potential workarounds, see some of the hidden comments on this thread. There are options for code configuration/massaging of the raw requests.

**Comment by nmarun** [comment]
I was able to get this working. I added 
`AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);`
to the constructor of my service proxy.
Thanks @OpenSpacesAndPlaces.

**Comment by OpenSpacesAndPlaces** [comment]
@nmarun

Whenever it's time for you to migrate - it looks like your problem may be fixed in 3.0?
https://github.com/dotnet/corefx/issues/32328
https://github.com/dotnet/corefx/pull/33426
https://github.com/dotnet/corefx/issues/32328#issuecomment-438301134
(also in-case this helps anybody else)

**Comment by flacidsnake** [comment]
Please, see #1257 
Hope this can help
